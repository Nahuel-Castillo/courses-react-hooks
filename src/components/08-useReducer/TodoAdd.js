import React from 'react';
import { useForm } from '../../hooks/useForm';

export const TodoAdd = ({ handleAddTodo }) => {

    const [ { description:inputValue }, handleInputChange, reset ] = useForm( { description: '' });

    const handleSubmit = e => {
        e.preventDefault();

        if ( inputValue.trim().length <= 1) return;

        const newTodo = {
            id: new Date().getTime(),
            desc: inputValue,
            done: false
        };

        handleAddTodo(newTodo);

        reset();
    }

    return (
        <>
            <h4> Add TODO </h4>
            <hr/>

            <form onSubmit={ handleSubmit }>

                <input 
                    type="text"
                    name="description"
                    placeholder="Learn..."
                    autoComplete="off"
                    className="form-control"
                    onChange={ handleInputChange }
                    value={ inputValue }
                />

                <button
                    type="submit"
                    className="btn btn-outline-success mt-1 btn-block"
                >
                    add
                </button>

            </form>
        </>
    )
}
