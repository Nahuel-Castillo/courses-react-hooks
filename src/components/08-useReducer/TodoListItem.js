
import React from 'react';
import PropTypes from 'prop-types';

export const TodoListItem = ( { todo, i, handleDelete, handleToggle } ) => {
    return (
        <li
            className="list-group-item"
        >
            <p
                className={`${todo.done && 'complete'}`}
                onClick={ () => handleToggle(todo.id) }
            >
                {i + 1}. {todo.desc}
            </p>

            <button
                className="btn btn-danger"
                onClick={() => handleDelete(todo.id)}
            >
                Remove
            </button>
        </li>
    )
}

TodoListItem.propTypes = {
    todo: PropTypes.object.isRequired,
    i: PropTypes.number.isRequired,
    handleDelete: PropTypes.func.isRequired,
    handleToggle: PropTypes.func.isRequired
}