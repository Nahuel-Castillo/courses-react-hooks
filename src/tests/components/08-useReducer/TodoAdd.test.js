import { shallow } from 'enzyme';
import { TodoAdd } from "../../../components/08-useReducer/TodoAdd";

describe('Test of <TodoAdd/>', () => {

    const handleAddTodo = jest.fn();

    const wrapper = shallow(<TodoAdd handleAddTodo={ handleAddTodo } />);

    test('should show correctly', () => {
        
        expect( wrapper ).toMatchSnapshot();
    });
   
    test('should not call handleAddTodo()', () => {
        
        const formSubmit = wrapper.find('form').prop('onSubmit');

        formSubmit( { preventDefault(){} } );

        expect( handleAddTodo ).not.toHaveBeenCalled();

    });

    test('should call handleAddTodo', () => {
        
        const value = 'hola';

        wrapper.find('input').simulate('change', { target: {
            name:'description',
            value
        }});

        const formSubmit = wrapper.find('form').prop('onSubmit');

        formSubmit( { preventDefault(){} } );

        expect( handleAddTodo ).toHaveBeenCalledWith( {
            id: expect.any(Number),
            desc: value,
            done: false
        });
    });
});