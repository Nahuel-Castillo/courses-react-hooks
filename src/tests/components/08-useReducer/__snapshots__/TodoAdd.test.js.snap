// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test of <TodoAdd/> should show correctly 1`] = `
<Fragment>
  <h4>
     Add TODO 
  </h4>
  <hr />
  <form
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="form-control"
      name="description"
      onChange={[Function]}
      placeholder="Learn..."
      type="text"
      value=""
    />
    <button
      className="btn btn-outline-success mt-1 btn-block"
      type="submit"
    >
      add
    </button>
  </form>
</Fragment>
`;
